File: ios/Reguerta/Reguerta/Domain/Shifts/ShiftSwapRequestRepository.swift
import Foundation

enum ShiftSwapCommand {
    case create(requestedShiftId: String, reason: String)
    case respond(requestId: String, candidateShiftId: String, response: ShiftSwapResponseStatus)
    case cancel(requestId: String)
    case apply(requestId: String, candidateShiftId: String)
}

enum ShiftSwapCommandError: Error, Equatable {
    case noCandidates
    case permissionDenied
    case conflict(code: String)
    case unavailable
    case invalidData
    case unknown
}

struct ShiftSwapTransitionResult: Equatable {
    let requestId: String
    let candidateCount: Int?
}

protocol ShiftSwapRequestRepository: Sendable {
    func allShiftSwapRequests(environment: SessionEnvironment) async throws -> [ShiftSwapRequest]
    func transition(
        _ command: ShiftSwapCommand,
        environment: SessionEnvironment
    ) async throws -> ShiftSwapTransitionResult
}

File: android/Reguerta/app/src/main/java/com/reguerta/user/domain/shifts/ShiftSwapRequestRepository.kt
package com.reguerta.user.domain.shifts

interface ShiftSwapRequestRepository {
    suspend fun getAllShiftSwapRequests(): List<ShiftSwapRequest>

    suspend fun createShiftSwapRequest(requestedShiftId: String, reason: String): String

    suspend fun respondToShiftSwapRequest(
        requestId: String,
        candidateShiftId: String,
        response: ShiftSwapResponseStatus,
    )

    suspend fun cancelShiftSwapRequest(requestId: String)

    suspend fun applyShiftSwapRequest(requestId: String, candidateShiftId: String)
}
